modules/sdk/src/main/scala/org/tessellation/sdk/domain/snapshot/services/GlobalL0Service.scala

```diff
@@ -37,20 +50,36 @@ object GlobalL0Service {
     l0GlobalSnapshotClient: L0GlobalSnapshotClient[F],
     globalL0ClusterStorage: L0ClusterStorage[F],
     lastGlobalSnapshotStorage: LastSnapshotStorage[F, GlobalIncrementalSnapshot, GlobalSnapshotInfo],
-    singlePullLimit: Option[PosLong]
+    singlePullLimit: Option[PosLong],
+    maybeMajorityPeerIdSet: Option[NonEmptySet[PeerId]]
   ): GlobalL0Service[F] =
     new GlobalL0Service[F] {
 
       private val logger = Slf4jLogger.getLogger[F]
+      private val maybeMajorityPeerIds = maybeMajorityPeerIdSet.map(_.toNonEmptyList)
 
       def pullLatestSnapshot: F[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)] =
         globalL0ClusterStorage.getRandomPeer.flatMap { l0Peer =>
           l0GlobalSnapshotClient.getLatest(l0Peer).flatMap {
-            case ((snapshot, state)) =>
+            case (snapshot, state) =>
               snapshot.toHashedWithSignatureCheck.flatMap(_.liftTo[F]).map((_, state))
           }
         }
 
+      def pullLatestSnapshotFromMajority: F[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)] =
+        maybeMajorityPeerIds match {
+          case None => pullLatestSnapshot
+          case Some(majorityPeerIds) =>
+            getLatestMajoritySnapshotData(majorityPeerIds).flatMap { majoritySnapshotData =>
+              majoritySnapshotData.tailRecM { msd =>
+                pullLatestSnapshot.map {
+                  case (hs, info) if hs.ordinal === msd.ordinal && hs.hash === msd.hash => (hs, info).asRight[MajoritySnapshotData]
```

Comment:
I think there may be a race condition here:
 - you determine the majority hash
 - the nodes advance one ordinal
 - pullLatestSnapshot pulls latest snapshot indeed, but this snapshot is the +1 +2 and so on
 - in consequence this function may never end as the loop will run expecting that finally the matching snapshot will be downloaded but in reality it will only pull apart even more and this loop will never stop running

It's a tricky functionality because we want the snapshot with state right here and we only store these for the latest snapshot. One thing that comes to my mind that could solve this is: first download the latest, and then check if majority hash for the latest you downloaded matches the majority hash for that ordinal based on the majority peers. If not, download the latest randomly again and try to determine the majority again. I think that adding some checks after downloading the latest snapshot with state is ok as we don't actually require the guaranteed latest snapshot (which we can't actually achieve) but a snapshot and state with high probability of being the latest. We could discuss it more and try to find a better solution. Anyway code left the way it is right now would be prone to this race condition and could never terminate leaving nodes in stuck state.

Reply:
@TheMMaciek I think I see what you're saying but I suspect it's not likely to happen because all of the peers will have had to advance in the time it takes for us to determine the majority hash.

However, maybe we can tweak it so that instead of pulling latest snapshot we instead pull the snapshot with the ordinal we determined.

Reply:
@sdavidp 
> it's not likely to happen because all of the peers will have had to advance in the time it takes for us to determine the majority hash.

we've already seen such things that "should not happen" or it's "0.1% chance for this to happen" and it broke the cluster for a few months, so we're rather sensitive for such things now. if that's possible, please rework this to ensure it can finalise under all conditions (either with success or error)

Reply:
The problem with pulling the ordinal we determined is that the nodes only store the last snapshot along with corresponding state - you can't pull an arbitrary ordinal with a corresponding state but only the latest one on the given node.

I think the probability of the described scenario happening is actually not that low. The nodes advance in quite a synchronized manner - gossip is pretty performant in that matter. So there may be a situation that nodes are close to the end of the ongoing consensus they still return the hash of the last finished snapshot, then your node has a brief connection issue lagging the requests and responses, you determine the majority hash, but in the meantime the ongoing consensus finishes and the nodes return a newer snapshot. Even without an interruption on your nodes end it may be the case that some of the nodes advance, your nodes asks them for a latest snapshot it's newer than the one you're looking for, so you ask another one, again not matching - two loops already happened other nodes that originally still had the snapshot you wanted made it to advance and right now none of them hash it as a latest one.

I would think Murphy's law here, if it can happen it will. If we already know that there is a case like this possible, it would be better to fix it right now and not have us scratch our heads couple months from now trying to figure out why are some nodes getting stuck from time to time.

Reply:
I think we should fix this. This is actually the same discussion we had on the majority peer selection with regards to ordinal, and some of the same logic applies

>However, maybe we can tweak it so that instead of pulling latest snapshot we instead pull the snapshot with the ordinal we determined.

Yes, however we invoke this we should determine the majority ordinal first, before attempting to determine majority hash, especially for small number of node case determining ordinal first is most important. Even if we determined the ordinal from the random node, it would also work (almost as well.) I imagine the easier way would first be determine majority latest ordinal from priority then majority latest hash from priority.

Reply:
☝️ That is what the majority function was doing.